zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/ElasticsearchExporter.java

```diff
@@ -0,0 +1,156 @@
+    // TODO validate configuration
+    context.setFilter(new ElasticsearchRecordFilter());
+    LOG.debug("Exporter configured with {}", configuration);
+  }
+
+  @Override
+  public void open(final Controller controller) {
+    this.controller = controller;
+    client = createClient();
+    // TODO createSchemaWhenNeeded();
+    writer = createBatchWriter();
+
+    scheduleDelayedFlush();
+
+    LOG.info("Exporter opened");
+  }
+
+  @Override
+  public void close() {
+    try {
+      flush();
+      updateLastExportedPosition();
+    } catch (final Exception e) {
+      LOG.warn("Failed to flush records before closing exporter.", e);
+    }
+
+    try {
+      client._transport().close();
+    } catch (final Exception e) {
+      LOG.warn("Failed to close elasticsearch client", e);
+    }
+
+    LOG.info("Exporter closed");
+  }
+
+  @Override
+  public void export(final Record<?> record) {
+    writer.addRecord(record);
+    lastPosition = record.getPosition();
+    if (shouldFlush()) {
+      flush();
+      // Update the record counters only after the flush was successful. If the synchronous flush
+      // fails then the exporter will be invoked with the same record again.
+      updateLastExportedPosition();
+    }
+  }
+
+  private static ElasticsearchClient createClient() {
+    // FIXME call a client factory to instantiate `client` here
+    return new ElasticsearchClient(null);
+  }
+
+  private boolean shouldFlush() {
+    // FIXME should compare against both batch size and memory limit
+    return writer.getBatchSize() >= configuration.bulk.size;
+  }
+
+  private ExporterBatchWriter createBatchWriter() {
+    // TODO register all handlers here
+    return ExporterBatchWriter.Builder.begin()
+        .withHandler(new UserRecordValueExportHandler())
+        .build();
+  }
+
+  private void scheduleDelayedFlush() {
+    controller.scheduleCancellableTask(
+        Duration.ofSeconds(configuration.bulk.delay), this::flushAndReschedule);
+  }
+
+  private void flushAndReschedule() {
+    try {
+      flush();
+      updateLastExportedPosition();
+    } catch (final Exception e) {
+      LOG.warn("Unexpected exception occurred on periodically flushing bulk, will retry later.", e);
+    }
+    scheduleDelayedFlush();
+  }
+
+  private void flush() {
+    try {
+      // TODO revisit the need to pass the BulkRequestBuilder and the ElasticsearchScriptBuilder as
+      // params here
+      final ElasticsearchBatchRequest batchRequest =
+          new ElasticsearchBatchRequest(
+              client, new BulkRequest.Builder(), new ElasticsearchScriptBuilder());
+      writer.flush(batchRequest);
+    } catch (final PersistenceException ex) {
+      throw new ElasticsearchExporterException(ex.getMessage(), ex);
+    }
+  }
+
+  private void updateLastExportedPosition() {
+    controller.updateLastExportedRecordPosition(lastPosition, null);
+  }
+
+  private record ElasticsearchRecordFilter() implements RecordFilter {
```

Comment:
It's not obvious to me if the supported `RecordTypes` and `ValueTypes` will need to be configurable at some point